File: javascript/211.Design-Add-and-Search-Words-Data-Structure.js
/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

class WordDictionary {
  constructor() {
    this.root = {};
  }

  /**
   * @param {string} word
   * @return {void}
   */
  addWord(word) {
    let node = this.root;

    for (const c of word) {
      if (!node[c]) {
        node[c] = {};
      }

      node = node[c];
    }

    node.isEnd = true;
  }

  /**
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    const dfs = (index, node) => {
      if (index === word.length) {
        return node.isEnd === true;
      }

      const c = word[index];

      if (c === ".") {
        for (const key in node) {
          if (key !== "isEnd" && dfs(index + 1, node[key])) {
            return true;
          }
        }

        return false;
      }

      if (!node[c]) {
        return false;
      }

      return dfs(index + 1, node[c]);
    };

    return dfs(0, this.root);
  }
}

File: python/211.Design-Add-and-Search-Words-Data-Structure.py
class WordDictionary:

    def __init__(self):
        self.root = {}

    def addWord(self, word: str) -> None:
        node = self.root

        for c in word:
            if c not in node:
                node[c] = {}
            node = node[c]

        node["isEnd"] = True

    def search(self, word: str) -> bool:
        def dfs(node, index):
            if index == len(word):
                return node.get("isEnd", False)

            c = word[index]

            if c == ".":
                for key in node:
                    if key != "isEnd":
                        if dfs(node[key], index + 1):
                            return True
                return False

            if c not in node:
                return False

            return dfs(node[c], index + 1)

        return dfs(self.root, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)